File: python/day08-oop-fundamentals/solutions.py
"""Day 8 reference solutions."""


class BankAccount:
    def __init__(self, owner, balance=0):
        self.owner = owner
        self.balance = balance

    def deposit(self, amount):
        self.balance += amount

    def withdraw(self, amount):
        if amount > self.balance:
            return False
        self.balance -= amount
        return True

    def __str__(self):
        return f"{self.owner}'s account: ${self.balance}"


class Rectangle:
    def __init__(self, width, height):
        self.width = width
        self.height = height

    def area(self):
        return self.width * self.height

    def perimeter(self):
        return 2 * (self.width + self.height)

    def is_square(self):
        return self.width == self.height


class Playlist:
    def __init__(self):
        self.songs = []

    def add_song(self, title):
        self.songs.append(title)

    def __len__(self):
        return len(self.songs)

File: typescript-javascript/day08-oop-fundamentals/solutions.js
// Day 8 reference solutions.

class BankAccount {
    #balance;

    constructor(owner, balance = 0) {
        this.owner = owner;
        this.#balance = balance;
    }

    deposit(amount) {
        this.#balance += amount;
    }

    withdraw(amount) {
        if (amount > this.#balance) {
            return false;
        }
        this.#balance -= amount;
        return true;
    }

    get balance() {
        return this.#balance;
    }
}

class Rectangle {
    constructor(width, height) {
        this.width = width;
        this.height = height;
    }

    area() {
        return this.width * this.height;
    }

    perimeter() {
        return 2 * (this.width + this.height);
    }

    isSquare() {
        return this.width === this.height;
    }
}

class Playlist {
    constructor() {
        this.songs = [];
    }

    addSong(title) {
        this.songs.push(title);
    }

    get length() {
        return this.songs.length;
    }
}

module.exports = { BankAccount, Rectangle, Playlist };
